### PR TITLE
Enemies and hitboxes now interact properly.

### PR DESCRIPTION
**Pull request by ngcthao**
Refactored and edited HitBox and Enemy entities. Enemies can now take damage and be destroyed.

Done--
- If an enemy crosses a hitbox, it will take damage every frame
- Enemies can only die once--previously could die twice before being removed.
- Currency counter is updated when enemy is killed by an ally unit

Not Done--
- Attack speed has not been accounted for yet.
- Enemy waves/levels have not been made yet.
- Enemy paths are not created yet.

### CODE REVIEW

File: src/js/renderables/allies/adult.js [hunk 1]
```diff
@@ -6,25 +6,14 @@ class AdultEntity extends AllyEntity {
 
     constructor(x, y, settings) {
```
Comment (dantran16) [inline src/js/renderables/allies/adult.js:7]: If you are updating line 9 to current change, remove settings from constructor and update anything that is called new Adult().

File: src/js/renderables/allies/child.js [hunk 2]
```diff
@@ -6,25 +6,14 @@ class ChildEntity extends AllyEntity {
 
     constructor(x, y, settings) {
         // call the parent constructor
-        super(x, y, settings);
+        super(x, y, {image: "white-dot", width: 25, height: 25});
```
Comment (dantran16) [inline src/js/renderables/allies/child.js:9]: Same as comment from foodie

File: src/js/renderables/allies/foodie.js [hunk 3]
```diff
@@ -6,25 +6,14 @@ class FoodieEntity extends AllyEntity {
 
     constructor(x, y, settings) {
         // call the parent constructor
-        super(x, y, settings);
```
Comment (dantran16) [inline src/js/renderables/allies/foodie.js:9]: Same as comment for foodie

File: src/js/renderables/enemies/enemy.js [hunk 4]
```diff
@@ -94,76 +84,56 @@ class Enemy extends me.Entity {
...
     }
 
     // Method to handle enemy death
     die() {
-        if(this.health <= 0) {
-
-        // Remove the enemy from the game world when it dies
-            this.onDestroy()
-
+        if(this.alive) {
```
Comment (dantran16) [inline src/js/renderables/enemies/enemy.js:97]: We don't need this.alive. Health should be enough to cause a death and isn't initialized anywhere.

File: src/js/renderables/enemies/enemy.js [hunk 5]
```diff
@@ -94,76 +84,56 @@ class Enemy extends me.Entity {
...
         console.log(`Player rewarded with ${this.reward} coins.`);
     }
 
+    // Remove the enemy from the world when it dies
+    onDestroy() {
+        if (this.alive) {
+            console.log(`${this} enemy is being removed from the game world.`);
+            me.game.world.removeChild(this); 
+        }
+        this.alive = false;
```
Comment (dantran16) [inline src/js/renderables/enemies/enemy.js:116]: Remove this.alive = false since if we destroy/remove it, we're referencing a null object.

Comment (ngcthao) [reply src/js/renderables/enemies/enemy.js:116]: this.alive is a built in attribute of melonJS's entity and has a default value of true--[Link](https://melonjs.github.io/melonJS/classes/Entity.html#alive)

I am not sure why, but I believe the removeChild method doesn't actually destroy the object, it just makes it not visible to the user. I can't find a method to free the resource (ie something like a destroy() method). So if I remove it, it causes the problem where the game will try to delete it multiple times.

**Review by dantran16 — APPROVED** [review]

**Pull request merged**